libs/server/Resp/Vector/VectorManager.Cleanup.cs

```diff
@@ -0,0 +1,386 @@
+                        {
+                            contextMetadata.FinishedCleaningUp(cleanedUp);
+                        }
+                    }
+
+                    UpdateContextMetadata(ref ctx);
+                }
+                catch (Exception e)
+                {
+                    logger?.LogError(e, "Failure during background cleanup of deleted vector sets, implies storage leak");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called in response to <see cref="TryMarkDeleteInProgress{TContext}(ref TContext, ref SpanByte, ulong)"/> or <see cref="ClearDeleteInProgress{TContext}(ref TContext, ref SpanByte, ulong)"/> to update metadata in Tsavorite.
+        /// </summary>
+        internal static void UpdateInProgressDeletes(Span<byte> updateMessage, ref SpanByte inLogValue, ref RecordInfo recordInfo, ref RMWInfo rmwInfo)
+        {
+            var context = BinaryPrimitives.ReadUInt64LittleEndian(updateMessage);
+            var len = BinaryPrimitives.ReadInt32LittleEndian(updateMessage[sizeof(ulong)..]);
+            var isAdding = len > 0;
+            var key = updateMessage[(sizeof(ulong) + sizeof(int))..];
+
+            Debug.Assert(key.Length == (isAdding ? len : -len), "Key length not expected");
+            Debug.Assert(context is >= ContextStep, "Special context not allowed");
+
+            var remaining = inLogValue.AsSpan();
+            while (remaining.Length >= sizeof(ulong) + sizeof(int))
+            {
+                var curCtx = BinaryPrimitives.ReadUInt64LittleEndian(remaining);
+
+                if (curCtx == 0)
+                {
+                    // Reached uninitialized data
+                    break;
+                }
+
+                var curLen = BinaryPrimitives.ReadInt32LittleEndian(remaining[sizeof(ulong)..]);
+                if (curCtx == context)
+                {
+                    if (isAdding)
+                    {
+                        // Already added, ignore and make no other changes
+                        return;
+                    }
+
+                    // Copy later values to cover the one we're removing
+                    var afterCur = remaining[(sizeof(ulong) + sizeof(int) + curLen)..];
+                    afterCur.CopyTo(remaining);
+
+                    // Clear everything after that so we won't think it's valid
+                    remaining[^(sizeof(ulong) + sizeof(int) + curLen)..].Clear();
+
+                    return;
+                }
+
+                remaining = remaining[(sizeof(ulong) + sizeof(int) + curLen)..];
+            }
+
+            if (isAdding)
+            {
+                // Not already added, so slap it in
+                BinaryPrimitives.WriteUInt64LittleEndian(remaining, context);
+                BinaryPrimitives.WriteInt32LittleEndian(remaining[sizeof(ulong)..], len);
+
+                key.CopyTo(remaining[(sizeof(ulong) + sizeof(int))..]);
+            }
```

Comment:
I saw in some other places where we have variable length data these functions calls:
```cs
                    rmwInfo.ClearExtraValueLength(ref recordInfo, ref value, value.TotalSize);
                    value.UnmarkExtraMetadata();
                    value.ShrinkSerializedLength(metadataSize + inputValue.Length + EtagConstants.EtagSize);
                    rmwInfo.SetUsedValueLength(ref recordInfo, ref value, value.TotalSize);
```
Do we need to manage the used length here as well? I added another comment asking about possibly unbounded lenght increase, maybe these would solve this issue? (if it is an issue)